hail/src/main/scala/is/hail/types/physical/PCanonicalShuffle.scala

```diff
@@ -0,0 +1,76 @@
+package is.hail.types.physical
+
+import is.hail.services.shuffler.Wire
+import is.hail.types.virtual._
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.expr.ir._
+import is.hail.utils._
+
+final case class PCanonicalShuffle(
+  val tShuffle: TShuffle,
+  required: Boolean = false
+) extends PShuffle {
+  def _asIdent = "shuffle"
+
+  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean): Unit = sb.append("PCShuffle")
+
+  val representation: PCanonicalBinary = PCanonicalBinary(required)
+
+  def codeOrdering(mb: EmitMethodBuilder[_], other: PType): CodeOrdering = {
+    assert(other isOfType this)
+    representation.codeOrdering(mb)
+  }
+
+  def setRequired(required: Boolean) = if(required == this.required) this else PCanonicalShuffle(tShuffle, required)
+
+  def unsafeOrdering(): UnsafeOrdering = representation.unsafeOrdering()
+}
+
+object PCanonicalShuffleSettable {
+  def apply(sb: SettableBuilder, pt: PCanonicalShuffle, name: String): PCanonicalShuffleSettable =
+    new PCanonicalShuffleSettable(pt, sb.newSettable[Long](s"${ name }_shuffle"))
+
+  def fromArrayBytes(cb: EmitCodeBuilder, region: Value[Region], pt: PCanonicalShuffle, bytes: Code[Array[Byte]]) = {
+    val off = cb.newField[Long](
+      "PCanonicalShuffleSettableOff",
+      pt.representation.allocate(region, Wire.ID_SIZE))
+    cb.append(pt.representation.store(off, bytes))
+    new PCanonicalShuffleSettable(pt, off)
+  }
+}
+
+class PCanonicalShuffleSettable(val pt: PCanonicalShuffle, shuffle: Settable[Long]) extends PShuffleValue with PSettable {
```

Comment:
can this class wrap a PCanonicalBinarySettable instead of a Settable[Long]? It makes me uncomfortable that we could break this if we changed PCanonicalBinary.

Reply:
done